.github/workflows/tests.yaml

```diff
@@ -23,19 +23,26 @@ jobs:
         # At present, these images are newer than "latest". We use them to test
         # against more recent tooling versions.
         # https://github.com/actions/runner-images
-        os: [ubuntu-22.04, macos-12]
+        os: [ubuntu-22.04, macos-14]
         build_mode: [fastbuild, opt]
-        # The clang-tidy config doesn't work on macos (missing `truncate`).
         include:
+          # The clang-tidy config doesn't work on macos (missing `truncate`).
           - os: 'ubuntu-22.04'
             build_mode: 'clang-tidy'
+          # On push (post-submit), also run on `macos-12` to get Intel macOS
+          # coverage. This is less and less critical so we skip it on the PR and
+          # merge queue path by switching to a redundant OS.
+          - os: ${{ github.event_name == 'push' && 'macos-12' || 'macos-14' }}
+            build_mode: fastbuild
+          - os: ${{ github.event_name == 'push' && 'macos-12' || 'macos-14' }}
+            build_mode: opt
     runs-on: ${{ matrix.os }}
 
     steps:
       # Ubuntu images start with 23GB available, and this adds 14GB more. For
       # comparison, MacOS images have >100GB free.
       - name: Free up disk space (Ubuntu)
-        if: matrix.os == 'ubuntu-22.04'
+        if: contains(matrix.os, 'ubuntu')
```

Comment:
Why contains instead of startsWith?

Reply:
It was shorter, but maybe that's not a great reason.

That said, this still was really repetitive. I've extracted a job-level environment context variable `os` to just contain `macos` or `ubuntu`, and renamed the matrix variable to `runner`. This lets us simply use `runner` when we want a *specific* runner image, and `os` when we just need to know whether its macOS or Ubuntu.

PTAL and let me know what you think. Happy to go back or try different approaches here.